airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/credential/S3AssumeRoleCredentialConfig.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+package io.airbyte.cdk.integrations.destination.s3.credential
+
+import com.amazonaws.auth.AWSCredentialsProvider
+import com.amazonaws.auth.STSAssumeRoleSessionCredentialsProvider
+import com.amazonaws.regions.Regions
+import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceClient
+
+private const val AIRBYTE_STS_SESSION_NAME = "airbyte-sts-session"
+
+/**
+ * The S3AssumeRoleCredentialConfig implementation of the S3CredentialConfig returns an
+ * STSAssumeRoleSessionCredentialsProvider. The STSAssumeRoleSessionCredentialsProvider
+ * automatically refreshes assumed role credentials on a background thread. To do this, an
+ * STS Client is created using the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables
+ * that are provided by the orchestrator. The roleArn comes from the spec and the externalId, which
+ * is used to protect against confused deputy problems, and also is provided through the
+ * orchestrator via an environment variable. As of 5/2024, the externalId is set to the workspaceId.
+ *
+ * @param roleArn The Amazon Resource Name (ARN) of the role to assume.
+ */
+class S3AssumeRoleCredentialConfig(private val roleArn: String) : S3CredentialConfig {
+    // TODO: Verify this env var, I think it might actually be AWS_ASSUME_ROLE_EXTERNAL_ID or something like that.
+    private val externalId: String? = System.getenv("AWS_EXTERNAL_ID")
```

Comment:
should that be in the config?

Reply:
This has to be set by us to a unique value for each customer. It cannot come from the config because that would allow for a confused deputy problem. This coming from the environment solves this problem because it means that even if you were to use an ARN that you don't own, the externalId will never match the customer's Role rules as the externalId will not match.

For example:

Customer A has configured a role with id: 1234 and externalId: abcd
Customer B has configured a role with id: 5678 and externalId: zxyw

If Customer A sets an ARN of 5678, the platform will still inject abcd as the externalId, so the STS Assume Role will fail because the externalId was not zxyw. The key here is that Airbyte MUST provide the externalId to the connector and the externalId MUST NOT be configurable.

 It is not required that the externalId (or the role ARN) be secret.
 
 More info here: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html#external-id-purpose